Add TitleView to VisualChildren list for Shell

drasticactions opened the pull request:
Fixes #13992

This adds TitleView into the Visual Tree list, allowing it to be hot reloaded. For this to work, I added IVisualTreeElement to Page and Shell, then overrode the list to include the TitleView, which is an View and automatically has its own children.

https://user-images.githubusercontent.com/898335/225862609-c084062a-7e84-4c37-9e6f-cc56aa3454c4.mov

drasticactions committed:
Add TitleView to GetVisualChildren

drasticactions committed:
Fix OnChildAdd/Remove

drasticactions committed:
Cast to page, rather than shell

PureWeen commented on src/Controls/src/Core/Shell/Shell.cs:
```diff
@@ -1316,10 +1326,16 @@ static void OnTitleViewChanged(BindableObject bindable, object oldValue, object
 			if (newView != null)
+			{
 				newView.Parent = owner;
+				VisualDiagnostics.OnChildAdded((Page)bindable, newView);
```
Can we just call AddLogicalChild here?

drasticactions replied:
Add/RemoveLogicalChild are on Shell (https://github.com/dotnet/maui/blob/9aa30fc98ebf8b6bd24d8f72c5724baae7f98145/src/Controls/src/Core/Shell/Shell.cs#L707) but `Bindable` could be a ContentPage (At least going by the unit tests that failed when I did cast to Shell (https://github.com/dotnet/maui/pull/14006/commits/fc2775f5caea72ef87d3ad9b0572b416246205e9)).

Seeing that this is in Shell, I could check if Bindable is Shell, and if so, call Add/RemoveLogicalChild, but if it's anything else it would be missed. Is that okay?

PureWeen replied:
I think so :-) 

There's two places we need to account for the TitleView being added

1) On the ContentPage
2) On the Shell

If it's added on the ContentPage then we want the `titleView` to get added to the LogicalChildren of the PAge not shell. If it's added to Shell we want it to get added to the LogicalChildren of Shell not Page

PureWeen commented on src/Controls/src/Core/Shell/Shell.cs:
```diff
@@ -1316,10 +1326,16 @@ static void OnTitleViewChanged(BindableObject bindable, object oldValue, object
 
 			if (oldView != null)
+			{
+				VisualDiagnostics.OnChildRemoved((Page)bindable, oldView, ((IVisualTreeElement)bindable).GetVisualChildren().IndexOf(oldView));
```
Can we just call RemoveLogicalChild here?

drasticactions committed:
Merge branch 'main' into dev/timill/page-children

drasticactions committed:
Use RemoveLogicalChild/AddLogicalChild

drasticactions commented:
I switched Shell to Add/RemoveLogicalChild and that seems to work fine. Page is proving a bit tougher. It has "LogicalChildren", "InternalLogicalChildren" and "InternalChildren". InternalChildren is public, but if I try and add TitleView to it, it seems to get stomped later on by TemplateUtilities.OnContentChanged, where it gets removed and replaced by the actual page content.

I think it's possible to refactor all of this down so that "Add/RemoveLogicalChild" is put into "Page" and out of Shell, so all of them can use it, but it'll take a bit more time.

drasticactions committed:
Revert changes to Page

drasticactions committed:
Nit

drasticactions commented:
@PureWeen I removed the `Page` code for now and I want to spin that into another PR. I think it will have more discussion around it, and as you're already doing stuff with InternalChildren (https://github.com/dotnet/maui/pull/14132) we could talk about the views that should appear inside InternalChildren and the visual tree. 

The Shell code is independent of it and switching it to use Remove/AddLogicalChild should have that work with whatever new thing is created, so that feels safe to add it as is right now.

PureWeen reviewed (changes requested):
Looks good! Can we add some tests for this inside Controls.Core.UnitTests?

drasticactions committed:
Merge branch 'main' into dev/timill/page-children

drasticactions committed:
Add Unit Test

drasticactions commented:
> Looks good! Can we add some tests for this inside Controls.Core.UnitTests?

Added with https://github.com/dotnet/maui/pull/14006/commits/658d2981ba82b7953fb809fa442ef08f8d76c936, is this okay?

drasticactions committed:
Should probably test Shell here, since that's what I fixed.
